runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcCommitWorkStream.java

```diff
@@ -204,56 +224,64 @@ private void flushInternal(Map<Long, PendingRequest> requests) {
     StreamingCommitWorkRequest chunk = requestBuilder.build();
-    synchronized (this) {
-      pending.put(id, pendingRequest);
-      try {
-        send(chunk);
-      } catch (IllegalStateException e) {
-        // Stream was broken, request will be retried when stream is reopened.
-      }
+    try {
+      send(chunk);
+    } catch (IllegalStateException e) {
+      // Stream was broken, request will be retried when stream is reopened.
     }
   }
 
   private void issueBatchedRequest(Map<Long, PendingRequest> requests) {
+    if (!prepareForSend(requests)) {
+      requests.forEach((ignored, pendingRequest) -> pendingRequest.abort());
+      return;
+    }
+
     StreamingCommitWorkRequest.Builder requestBuilder = StreamingCommitWorkRequest.newBuilder();
     String lastComputation = null;
     for (Map.Entry<Long, PendingRequest> entry : requests.entrySet()) {
       PendingRequest request = entry.getValue();
       StreamingCommitRequestChunk.Builder chunkBuilder = requestBuilder.addCommitChunkBuilder();
-      if (lastComputation == null || !lastComputation.equals(request.computation)) {
-        chunkBuilder.setComputationId(request.computation);
-        lastComputation = request.computation;
+      if (lastComputation == null || !lastComputation.equals(request.computationId())) {
+        chunkBuilder.setComputationId(request.computationId());
+        lastComputation = request.computationId();
       }
-      chunkBuilder.setRequestId(entry.getKey());
-      chunkBuilder.setShardingKey(request.request.getShardingKey());
-      chunkBuilder.setSerializedWorkItemCommit(request.request.toByteString());
+      chunkBuilder
+          .setRequestId(entry.getKey())
+          .setShardingKey(request.shardingKey())
+          .setSerializedWorkItemCommit(request.serializedCommit());
     }
     StreamingCommitWorkRequest request = requestBuilder.build();
-    synchronized (this) {
-      pending.putAll(requests);
-      try {
-        send(request);
-      } catch (IllegalStateException e) {
-        // Stream was broken, request will be retried when stream is reopened.
-      }
+    try {
+      send(request);
+    } catch (IllegalStateException e) {
+      // Stream was broken, request will be retried when stream is reopened.
     }
   }
 
-  private void issueMultiChunkRequest(final long id, PendingRequest pendingRequest) {
-    checkNotNull(pendingRequest.computation);
-    final ByteString serializedCommit = pendingRequest.request.toByteString();
+  private void issueMultiChunkRequest(long id, PendingRequest pendingRequest) {
+    if (!prepareForSend(id, pendingRequest)) {
+      pendingRequest.abort();
+      return;
+    }
 
+    checkNotNull(pendingRequest.computationId(), "Cannot commit WorkItem w/o a computationId.");
+    ByteString serializedCommit = pendingRequest.serializedCommit();
     synchronized (this) {
-      pending.put(id, pendingRequest);
       for (int i = 0;
```

Comment:
moving the adding to pending requests out of this synchronized block means that it is possible for the stream to be recreated and the items already removed from pendingRequests.  Then we will end up sending the requests multiple  times on the new stream and we may get multiple responses unexpectedly.

Can prepareforsend be called beneath the synchronized block to minimize locking changes?

Reply:
done